### ISSUE TITLE
ESLint might give wrong position information

### ISSUE THREAD

**Issue by liuliangsir** [eslint/eslint#13214]
As the above title says, It's because that ESLint give wrong position information.

**Tell us about your environment**
- ESLint Version: 6.8.0
- Node Version: 10.15.3
- npm Version: 6.4.1
- OS System: Windows 10 Pro(64-bit)

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration**:
[Here](https://gist.github.com/liuliangsir/5b69133e73577576771ff127947940b6) is a configuration file for ESLint

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint**.
[Here](https://github.com/liuliangsir/codemirror-lint-eslint) is a minimal reproducible example, And the following is some code snippet

```javascript
import isNil from 'lodash/isNil'
import linterModule from 'eslint/lib/linter/linter'

import defaultEslintConfig from '../assets/json/eslint.json'

const Pos = CodeMirror.Pos
const linter = new linterModule.Linter()

function validator(text, options) {
  const result = []
  if (!text) return result

  const [
    config = defaultEslintConfig,
    wrappedText = `async () => {\n${text}\n}`,
  ] = []
  const errors = linter.verify(wrappedText, {
    ...config,
  })
  for (let i = 0; i < errors.length; i++) {
    const error = errors[i]

    result.push({
      message: error.message,
      severity: getSeverity(error),
      from: Pos(error.line - 1 - 1, error.column - 1),
      to: Pos(
        isNil(error.endLine) ? error.line - 1 - 1 : error.endLine - 1 - 1,
        isNil(error.endColumn) ? error.column : error.endColumn - 1,
      ),
    })
  }

  return result
}

CodeMirror.registerHelper('lint', 'javascript', validator)

function getSeverity(error) {
  switch (error.severity) {
    case 1:
      return 'warning'
    case 2:
    default:
      return 'error'
  }
}
```
The command that I used to run ESLint is `npm start`
```bash
npm start
```

**What did you expect to happen**?
Showing lint message after removing the right-most close curly

![CodeMirror_-Linter-Demo-Google-Chrome-2020-04-23-16-10-57](https://user-images.githubusercontent.com/8049878/80075691-ac849e00-857d-11ea-9151-9201bd62c010.gif)

**What actually happened? Please include the actual, raw output from ESLint**.
Not showing lint message after removing the right-most close curly

![CodeMirror_-Linter-Demo-Google-Chrome-2020-04-22-14-58-44](https://user-images.githubusercontent.com/8049878/79951256-f05aa300-84aa-11ea-8d54-6a9fff70cd03.gif)

**Are you willing to submit a pull request to fix this bug**?
No

**Related Issue**
https://github.com/codemirror/CodeMirror/issues/6240

**Comment by anikethsaha** [comment]
I can't reproduce this in the demo. as it does shows parsing error when you remove the right-most curly brace. I did try with the wrapped text and its working fine.

[demo](https://eslint.org/demo#eyJ0ZXh0IjoiYXN5bmMgKCkgPT4ge1xuICBmdW5jdGlvbiBmb28oKSB7XG59Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoxMSwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==)

I think you are getting the correct error but it's being hidden because of the wrapped code.
for the following code, 

```js
function () {}
```

as you are wrapping the test, the following code is being sent to eslint

```js
async () => { // this might be hidden from the codemirror playground
function () {}
}  // this is too hidden 
```

so when you remove the right-most curly braces of the code in the playground, its actually removing the braces of inner function which is visible in the playground.
so the code which is sent is this

```js
async () => { 
function () { // you removed this `}` brace
}  
```
now the eslint will report a error with unexpected token at line 3, cause the inner function is fine ,  there is missing `}` for the async function 

```js
async () => { 
function () { // no error here
}   // 3:2 - Parsing error: Unexpected token
```

but this is not visible in the playground so that's why **I think** its not showing you the error here


I am not sure about this as I don't know-how `codemirror-lint-eslint` works under the hood but it seems eslint is working fine. 

***

Also, the missing highlighted position in the eslint demo page might be cause it doesn't highlight missing tokens in the playground. but the position is correct 👍

**Comment by mdjermanovic** [comment]
@liuliangsir thanks for the issue!

I believe @anikethsaha's analysis is correct, so you'll probably need to adjust the code that transforms positions from the wrapped input to the original input.

Nevertheless, there's not enough information to evaluate if this is really a bug with reported locations. Could you please make a file with the code in question, then run eslint with `-f json`  on that file, and post both the code and the output here?

**Comment by liuliangsir** [comment]
@mdjermanovic using the awkward workaround is a sad thing. Check out https://github.com/babel/babel-eslint/issues/829 and https://github.com/codemirror/CodeMirror/issues/6240 for more details.

Here is the code
```javascript
async () => {
  function bar() {
}
```
And here is the output
```json
[{"filePath":"D:\\work\\github\\codemirror-lint-eslint\\src\\lint.js","messages":[{"ruleId":null,"fatal":true,"severity":2,"message":"Parsing error: Unexpected token","line":3,"column":2}],"errorCount":1,"warningCount":0,"fixableErrorCount":0,"fixableWarningCount":0,"source":"async () => {\n  function bar() {\n}"}]
```

**Comment by mdjermanovic** [comment]
Thanks for the additional info!

I also couldn't think of a better workaround for top-level await than the one you found.

As for the position issue, `(3:2)` seems like a valid location in the wrapped input.

If it helps, you can use [SourceCode#splitLines()](https://eslint.org/docs/developer-guide/nodejs-api#sourcecode-splitlines), or `linter.getSourceCode()` (please see the last section of [Linter#verify](https://eslint.org/docs/developer-guide/nodejs-api#linter-verify)) which returns a [SourceCode](https://eslint.org/docs/developer-guide/working-with-rules#context-getsourcecode) object with `lines` property. Then you could find out if the location is out of bounds for the original input and adjust it.
